video: add stm32mp135f gc2145 camera board shield & enable dcmipp / st-mipid02

avolmat-st opened the pull request:
This PR adds the shield for the GC2145 camera module available with the stm32mp135f-dk

This PR depends on PR #91878 for allowing crop in order to properly display the image

Commit:
boards: st: stm32mp135f_dk: add camera pipeline nodes

The STM32MP135F-DK board embeds a ST-MIPID02 CSI to DVP
bridge allowing to connect a CSI 15pins FFC cable enabled
sensor to the DCMIPP in parallel interface.
Add the ST-MIPID02 node and its dependency MCO clock as well
as partially filled DCMIPP node.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
shields: st_mb1897_cam: addition of MB1897 camera board

Addition of a shield st_mb1897_cam embedding a GC2145 CSI
sensor which can be connected to the STM32MP135F-DK via a 15pins FFC.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: video: capture: add a stm32mp135f_dk config

Add a configuration file for the stm32mp135f_dk board.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

erwango reviewed (changes requested):
Otherwise, looks ok.

josuah commented on boards/st/stm32mp135f_dk/Kconfig.defconfig:
```diff
@@ -0,0 +1,15 @@
+
+if VIDEO
+
+config VIDEO_INIT_PRIORITY
+	default 75
```
[question]
Currently, video drivers have default priority of [`60`](https://github.com/zephyrproject-rtos/zephyr/blob/10fe120488bd20203782beebf8a0171247af6c8b/drivers/video/Kconfig#L22) which cover all image sensors. Drivers that need to be initialized after pick [`61`](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/Kconfig.mcux_mipi_csi2rx#L14), and drivers that need to be initialized before pick [`59`](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/Kconfig.mcux_csi#L14)

Setting `VIDEO_INIT_PRIORITYY` to `75` would be switching the image sensors instead of switching the individual drivers that need initialization before or after. Would it work to modify the driver init priority instead?

If not, a comment indicating why this is needed can be great for maintaining things in the long run.

avolmat-st replied:
The reason why I set this to 75 was actually because

config GPIO_MCP230XX_INIT_PRIORITY
        int "MCP230XX GPIO expander init priority"
        default 75
        help
          Device driver initialization priority.

but indeed, I could maybe have changed instead the MCP230XX_INIT_PRIORITY instead to avoid changing the video init priority. Let me check that.

josuah reviewed (commented):
LGTM thank you!

Just waiting to get your feedback on init priorities to confirm this was intentional, but I stand ready to give a quick review cycle to approve after that is clarified.

Commit:
board: st: stm32mp135f_dk: align MCP230XX init prio to I2C

MCP230XX IO expander is interfaced via i2c hence it is
suffisant to align its init priority to the i2c one.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah reviewed (dismissed):
LGTM!
I think CI will sort itself out once #91878 will be merged...

avolmat-st commented:
The compliance checks failure are due to the dependency on the PR #91878 since here I add a board specific sample capture app conf which include CROP Kconfig.

avolmat-st commented:
Update the DTSI / shield overlay following the proposal of a unified raspberrypi csi-connector in PR #91983

avolmat-st commented:
Updated to add video-interface.h include following @ngphibang review.

josuah commented on samples/drivers/video/capture/boards/stm32mp135f_dk.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_VIDEO_FRAME_WIDTH=480
+CONFIG_VIDEO_FRAME_HEIGHT=272
+CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=262000
+CONFIG_VIDEO_SOURCE_CROP_LEFT=80
+CONFIG_VIDEO_SOURCE_CROP_TOP=104
+CONFIG_VIDEO_SOURCE_CROP_WIDTH=480
+CONFIG_VIDEO_SOURCE_CROP_HEIGHT=272
```
I think these were bothering CI but let's wait dependencies to be merged, maybe no need to debug CI until then...

avolmat-st commented on boards/shields/st_stm32mp13_gc2145_cam/doc/index.rst:
```diff
@@ -0,0 +1,32 @@
+.. _st_stm32mp13_gc2145_cam:
+
+ST_STM32MP13_GC2145_CAM
```
Indeed, it can be used anyway, but it is sold bundled with the stm32mp135f-dk board and apparently cannot be purchased separately. That's why I kept that long name. But indeed, that's not a technical aspect.

What about full_name ? : ST MB1897 camera module
                   shield name ? : st_mb1897_cam

kartben replied:
> What about full_name ? : ST MB1897 camera module
> shield name ? : st_mb1897_cam

yeah that's perfect, IMO :)

josuah commented on boards/shields/st_stm32mp13_gc2145_cam/st_stm32mp13_gc2145_cam.overlay:
```diff
@@ -0,0 +1,41 @@
+	data-lanes = <1 2>;
+};
+
+&csi_i2c {
```
On other shields, I forgot to add a `status = "okay";` for the `i2c`. Maybe it helps in case I2C is not enabled by default?

avolmat-st replied:
Yep, good point. I add it. Thanks.

josuah reviewed (commented):
I will be around through the day and tomorrow focusing on reviews/modifications as needed.

One feedback for the `status = "okay";` waiting your opinion.

josuah commented on boards/st/stm32mp135f_dk/stm32mp135f_dk.dts:
```diff
@@ -124,10 +146,64 @@
+		};
+
+		port@1 {
+			csi_capture_port: endpoint { };
```
Apparently `dtc` does not complain that the `remote-endpoint-label` required property is absent?
Then no need to add the stub `remote-endpoint-label = "";` I suppose!

avolmat-st replied:
Not on that one since there is a compatible in the stm32mp135.dtsi

ngphibang replied:
I don't see where the `compatible = "st,stm32-dcmipp-pipe";` is. But I think inside that .yaml, we should still `include: video-interfaces.yaml` because this is an endpoint no ? If not, it seems we made an exception for this endpoint

ngphibang replied:
> That's how each pipe output is being described, based on the very first discussion we had at the beginning when I started implementing that.

Thanks for reminding me. Yes, I remember we have some constraints to support multiple pipes in the DCMIPP and come up to this solution but didn't pay attention to this particular trade-off.

This is needed but not a real endpoint (no remote-endpoint to interface with, no bus-type). Maybe we can try to see if we can have a nicer solution in the future. The current implementation is OK to me. Thanks.

josuah reviewed (dismissed):
Glad to see more shields being used, that reduces efforts of manually wiring things up in devicetree.

ngphibang reviewed (changes requested):
Added change-request to wait for more discussion about my question on `csi_capture_port: endpoint { };` maybe I am wrong, will remove it as soon as more info is provided

avolmat-st commented:
> Added change-request to wait for more discussion about my question on `csi_capture_port: endpoint { };` maybe I am wrong, will remove it as soon as more info is provided

This is part of the stm32mp135.dtsi in the description of the dcmipp node:

```
                dcmipp: dcmipp@5a000000 {
                        compatible = "st,stm32mp13-dcmipp", "st,stm32-dcmipp";
                        reg = <0x5a000000 0x400>;
                        interrupts = <GIC_SPI 79 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
                        clocks = <&rcc STM32_CLOCK(APB4, 1)>,
                                 <&rcc STM32_SRC_PLL2_Q DCMIPP_SEL(1)>;
                        clock-names = "dcmipp", "dcmipp-ker";
                        resets = <&rctl STM32_RESET(APB4, 1)>;
                        status = "disabled";

                        ports {
                                #address-cells = <1>;
                                #size-cells = <0>;

                                port@0 {
                                        reg = <0>;

                                        endpoint {
                                                remote-endpoint-label = "";
                                                bus-type = <VIDEO_BUS_TYPE_PARALLEL>;
                                        };
                                };

                                port@1 {
                                        reg = <1>;

                                        dcmipp_pipe_dump: endpoint {
                                                compatible = "st,stm32-dcmipp-pipe";
                                        };
                                };
                        };
                };

```

That's how each pipe output is being described, based on the very first discussion we had at the beginning when I started implementing that.

avolmat-st commented:
> adding 100KB to the repo for such a tiny image was really overkill :) updated to webp

Thanks @kartben, sorry I didn't noticed ;(
